src/util/system.h

```diff
@@ -501,6 +502,21 @@ inline void insert(std::set<TsetT>& dst, const Tsrc& src) {
     dst.insert(src.begin(), src.end());
 }
 
+/**
+ * Helper function to access the contained object of a std::any instance.
+ * Returns a pointer to the object if passed instance has a value and the type
+ * matches, nullptr otherwise.
+ */
+template<typename T>
+T* AnyPtr(const std::any& any)
```

Comment:
Concept NACK this implementation.

You should label this function as noexcept and use the noexcept versions of any_cast which will already return you a nullptr. See variants 4 and 5 https://en.cppreference.com/w/cpp/utility/any/any_cast. These are implemented AFAIK internally without using exceptions.

Exceptions have a non negligible performance overhead if hitting the `catch` case we may want to avoid.

You could -- if you want -- introduce a try/catch variant that is called `AnyPtrLikely` that uses try/catch where we're already pretty sure that we'll succeed, but that seems like over optimizing.

Reply:
Oops, sorry. I didn't know about the noexcept casts. This can be simplified to:

```c++
template<typename T>
T* AnyPtr(const std::any& any)
{
    T* const* ptr = std::any_cast<T*>(&any);
    return ptr ? *ptr : nullptr;
}
```

I don't think this is used in any performance critical code, but no reason this needs to be slower and more complicated.

Reply:
Good catch, I somehow missed that there was a noexcept version of `std::any_cast`. But then I guess we can as well directly use that instead of introducing AnyPtr? This would be in effect just another name for the same purpose.

Reply:
> Good catch, I somehow missed that there was a noexcept version of `std::any_cast`. But then I guess we can as well directly use that instead of introducing AnyPtr? This would be in effect just another name for the same purpose.

My preference would be to keep AnyPtr instead of introducing NodeContext* const* temporary variables, but feel free to take any approach.

Reply:
using any_cast bare is a code smell IMO because it can arbitrarily become the non noexcept version :(, the wrapper makes the version guaranteed. If anything, you could declare it as inline so there is no overhead? (idk about template + inline...)

I think that's what @ryanofsky is saying.

Reply:
After having tried out using std::any_cast directly yesterday, I can confirm ryanofsky's point of the drawback, needing temporary variables, making the code clunky and potentially confusing. Also the ambiguity on which version is chosen is a strong argument against it. Kept the AnyPtr function, changed it to the [proposed implementation](https://github.com/bitcoin/bitcoin/pull/21366#discussion_r588923880), see https://github.com/bitcoin/bitcoin/pull/21366#issuecomment-792124283.